src/client/opamArg.ml

```diff
@@ -782,62 +774,414 @@ let opamlist_columns =
+    let flags = List.map (fun (v,c,f,d) -> contented_validity v c, f, d) flags in
+    let info_flags =
+      List.map (fun (validity, flag, doc) ->
+          let doc = update_doc_w_cli doc ~cli validity in
+          validity.content, Arg.info ?docs:section flag ~doc)
+        flags
+    in
+    let check selected =
+      let neww, rm =
+        List.fold_left (fun (neww,rm) elem ->
+            match OpamStd.List.find_opt (fun (validity, _, _) ->
+                validity.content = elem) flags with
+            | Some ({ removed = None ; valid = c; _ }, flags, _)
+              when cond_new cli c ->
+              (flags, c)::neww, rm
+            | Some ({ removed = Some (removal, instead); _ }, flags, _)
+              when cond_removed cli removal ->
+              neww, (flags, (removal, instead))::rm
+            | _ -> neww,rm) ([],[]) selected
+      in
+      match neww, rm with
+      | [], [] -> `Ok (List.map elem_of_vr selected)
+      | [flags, c], [] ->
+        error_new_flag cli c flags
+      | [], [flags, (c, instead)] ->
+        error_rm_flag cli c instead flags
+      | _::_, []->
+        let options, clis = List.split neww in
+        let msg =
+          Printf.sprintf
+            "%s are only available since respectively versions %s of the opam \
+             CLI, but version %s has been requested."
+            (OpamStd.Format.pretty_list (List.map get_long_form options) )
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string clis))
+            (OpamCLIVersion.to_string cli)
+        in
+        `Error (false,msg)
+      | [], _::_->
+        let options, clis = List.split rm in
+        let clis = List.split clis |> fst in
+        let msg =
+          Printf.sprintf
+            "%s were removed in, respectively, %s versions of the opam CLI, \
+             but version %s has been requested."
+            (OpamStd.Format.pretty_list (List.map get_long_form options) )
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string clis))
+            (OpamCLIVersion.to_string cli)
+        in
+        `Error (false,msg)
+      | _,_ ->
+        let news, nclis = List.split neww in
+        let rms, rclis = List.split rm in
+        let rclis = List.split rclis |> fst in
+        let msg =
+          Printf.sprintf
+            "%s are only available since respectively versions %s of the opam \
+             CLI,and %s were removed in respectively versions %s, \
+             but version %s has been requested."
+            (OpamStd.Format.pretty_list (List.map get_long_form news))
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string nclis))
+            (OpamStd.Format.pretty_list (List.map get_long_form rms))
+            (OpamStd.Format.pretty_list
+               (List.map OpamCLIVersion.to_string rclis))
+            (OpamCLIVersion.to_string cli)
+        in
```

Comment:
I think this needs to be a little more complicated (sadly!): if _any_ of the options don't exist any more (i.e. have no alternate flag name) then this CLI isn't valid in any version and the message is:

```
            "This combination of options is not possible: %s require
             at least version %s of the opam CLI but %s were all \
             removed by version %s of the opam CLI!"
```
(possibly with the same "removed by"/"removed in" distinction, if you wanted). I wouldn't mention the requested version of the CLI - in this case, there's no version.

If all of `rms` were in fact renamed, then the message is:
```
            "This combination of options is not possible: %s require
             at least version %s of the opam CLI and the newer %s \
             flags must be used for %s respectively!"
```
where the last parameters are the list of old options and the list they were renamed (which I think is in `snd rclis`?)